tripy/tripy/function_registry.py

```diff
@@ -356,23 +356,54 @@ def raise_overload_error(msg, candidate_overloads, mismatch_reasons=None, extra_
 
         raise_overload_error("Could not find an implementation", self.overloads[key], mismatch_reasons)
 
-    def __call__(self, key: Any):
+    def __call__(self, key: Any, bypass_dispatch: Optional[Union[bool, Sequence[str]]] = None):
         """
         Registers a function with this function registry.
         This function allows instances of the class to be used as decorators.
 
         Args:
             key: The key under which to register the function.
+
+            bypass_dispatch: Has no effect if the value is None or False.
+               If the decorator is applied to a function and the value is true, this will add the decorated function to the registry but does not
+               attempt to dispatch overloads. This option avoids overhead but does not perform automatic type-checking or allow for choosing between overloads.
+               If the decorator is applied to a class and the value is true, dispatch will be bypassed for all methods.
+               If the decorator is applied to a class and the value is a list of method names, dispatch will be bypassed for the listed methods only.
         """
+        bypass_dispatch = bypass_dispatch or False
 
         def impl(func):
             # Cannot dispatch on properties, so we just use func directly.
             # Otherwise, we generate a dispatch function which will dispatch to the appropriate overload
             # based on argument types.
             if isinstance(func, property):
                 self[key] = func
+            # For classes, we apply the wrapper to all methods.
+            elif inspect.isclass(func):
+                # Ignore properties and functions not defined in the class (we will use the presence of a docstring as a proxy for that).
+                # It does not suffice to check just that the method is inherited because some decorators like @dataclass add methods
+                # that are not documented or annotated and do not use inheritance to do so.
+                for name, member in inspect.getmembers(
+                    func, predicate=lambda m: inspect.isfunction(m) and m.__doc__ is not None
+                ):
+                    bypass_func = bypass_dispatch if isinstance(bypass_dispatch, bool) else name in bypass_dispatch
+                    setattr(func, name, self(f"{key}.{name}", bypass_dispatch=bypass_func)(member))
+                self[key] = func
             else:
+                if bypass_dispatch and key in self:
+                    from tripy.common.exception import raise_error
+
+                    raise_error(
+                        f"Attempting to add key {key} into a function registry with dispatch disabled, but there is already an overload present."
+                    )
```

Comment:
This should be an `assert` since it's not a user-facing error.

Reply:
Good point.